File: src/net/tc/testsuite/TestBase.java
package net.tc.testsuite;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import net.tc.data.db.ConnectionProvider;
import net.tc.utils.Utility;

public class TestBase {
	ConnectionProvider connectionProvider = null;
	Map config = null;
	int loops = 50;
	int sleep = 0;
	boolean verbose = false;
	boolean summary = false;
	boolean printConnectionTime = true;
	static String defaultsConnection = "&url=jdbc:mysql://127.0.0.1:3306&user=test_user&password=test_password&schema=test";
	static String defaultParameters = "&useSSL=false&autoReconnect=true";
	Map parameters = new HashMap();
	ArrayList startConnTimes = new ArrayList();
	ArrayList endConnTimes = new ArrayList();
	int formattingLenghtForNano = 15;
	DecimalFormat df2 = new DecimalFormat("#,###,###,##0.00");
	DecimalFormat dfCSV = new DecimalFormat("#.00");
	boolean reportCSV = false;

	
	
	
	
	
	void executeLocalAction(TestBase test) {
		// Generate this in the local class 
		
	}


	void init (String[] args){
	
		try {
			/*
			 * Configuration first
			 */
			this.generateParameters(defaultParameters);
			if (args.length >= 1) {
				this.getConfig().putAll(this.getArguments(args));
//				this.setConfig(this.getArguments(args));
			}
		} catch (Throwable th) {
			th.printStackTrace();
		}
		
		
		
//		setConnectionProvider(new ConnectionProvider(getConfig()));
		
		
		this.setLoops((this.getConfig().get("loops")!=null)
				?Utility.isNumeric(this.getConfig().get("loops"))?(int)getConfig().get("loop")
						:Integer.parseInt((String)getConfig().get("loops")):50); 
		
		this.setSleep((this.getConfig().get("sleep")!=null)
				?Utility.isNumeric(this.getConfig().get("sleep"))?(int)getConfig().get("sleep")
						:Integer.parseInt((String)getConfig().get("sleep")):0); 
		
		
		this.setVerbose((this.getConfig().get("verbose")!=null)?Boolean.parseBoolean((String)getConfig().get("verbose")):false);
		this.setSummary((this.getConfig().get("summary")!=null)?Boolean.parseBoolean((String)getConfig().get("summary")):false);
		
		this.setReportCSV((this.getConfig().get("reportCSV")!=null)?Boolean.parseBoolean((String)getConfig().get("reportCSV")):false);
		
	}
	
	void localInit(String[] args) {
		/*
		 * Add in your class specific settings
		 */
	}
	
	
	Map<String, Object> getArguments(String[] args) {
		if (args == null || args.length < 1) {
			this.showHelp();
		}

		Map<String, Object> config = new HashMap();
		for (String entry : args) {

			if (entry.indexOf("parameters") > -1) {
				this.generateParameters(entry.replace("parameters=", ""));
			} else {
				String keyValue[] = entry.split("=");
				config.put(keyValue[0], keyValue[1]);
			}
		}
		config.put("parameters", this.getParameters());
		return config;
	}
 
	ConnectionProvider getConnectionProvider() {
		return connectionProvider;
	}

	void setConnectionProvider(ConnectionProvider connectionProvider) {
		this.connectionProvider = connectionProvider;
	}

	Map<String, Object> getConfig() {
		return config;
	}

	void setConfig(Map config) {
		this.config = config;
	}

	Map getParameters() {
		return parameters;
	}

	void generateParameters(String params) {

		String[] paraTem = params.split("&");
		for (String entry : paraTem) {
			if (entry.indexOf("=") > 0) {
				String[] keyValue = entry.split("=");
				getParameters().put(keyValue[0], keyValue[1]);
			}

		}

	}

	void generateConfig(String configString) {

		String[] temp = configString.split("&");
		for (String entry : temp) {
			if (entry.indexOf("=") > 0) {
				String[] keyValue = entry.split("=");
				getConfig().put(keyValue[0], keyValue[1]);
			}

		}

	}

	void setParameters(Map parameters) {
		this.parameters = parameters;
	}

	 StringBuffer showHelp() {

		StringBuffer sb = new StringBuffer();
		sb.append("******************************************\n");
		sb.append("DB Parameters to use\n");
		sb.append("Parameters are COMMA separated and the whole set must be pass as string\n");
		sb.append(
				"IE java -Xms2G -Xmx3G -classpath \"./*:./lib/*\" net.tc.testsuite.MySQLConnectionTest \"loops=10,parameters=&characterEncoding=UTF-8, url=jdbc:mysql://192.168.4.22:3306\" \n");
		sb.append("url [url=jdbc:mysql://127.0.0.1:3306]\n");
		sb.append("user [user=test_user]\n");
		sb.append("password [password=test_pw]\n");
		sb.append("parameters [parameters=&useSSL=false&autoReconnect=true]\n");
		sb.append("schema [schema=test]\n");
		sb.append("\n*****************************************\nApplication Parameters \n");
		sb.append("loops [loops=50\n");
		sb.append("sleep [sleep=0] value in milliseconds \n");
		sb.append("verbose [verbose=false]\n");
		sb.append("summary [summary=false]\n");

		sb.append("reportCSV [reportCSV=false]\n");

		sb.append("****************************************\n Optional ");
		sb.append("selectForceAutocommitOff [selectForceAutocommitOff=true]\n");
		return sb;
		
		//		System.out.print(sb.toString());
//		System.exit(0);

	}




	int getLoops() {
		return loops;
	}




	void setLoops(int loops) {
		this.loops = loops;
	}




	int getSleep() {
		return sleep;
	}




	void setSleep(int sleep) {
		this.sleep = sleep;
	}




	boolean isVerbose() {
		return verbose;
	}




	void setVerbose(boolean verbose) {
		this.verbose = verbose;
	}




	boolean isSummary() {
		return summary;
	}




	void setSummary(boolean summary) {
		this.summary = summary;
	}




	static String getDefaultsConnection() {
		return defaultsConnection;
	}




	static void setDefaultsConnection(String defaultsConnection) {
		TestBase.defaultsConnection = defaultsConnection;
	}




	static String getDefaultParameters() {
		return defaultParameters;
	}




	static void setDefaultParameters(String defaultParameters) {
		TestBase.defaultParameters = defaultParameters;
	}




	int getFormattingLenghtForNano() {
		return formattingLenghtForNano;
	}




	void setFormattingLenghtForNano(int formattingLenghtForNano) {
		this.formattingLenghtForNano = formattingLenghtForNano;
	}




	DecimalFormat getDf2() {
		return df2;
	}




	void setDf2(DecimalFormat df2) {
		this.df2 = df2;
	}




	DecimalFormat getDfCSV() {
		return dfCSV;
	}




	void setDfCSV(DecimalFormat dfCSV) {
		this.dfCSV = dfCSV;
	}




	boolean isReportCSV() {
		return reportCSV;
	}




	void setReportCSV(boolean reportCSV) {
		this.reportCSV = reportCSV;
	}

}

File: src/modules/TestBase.go
package modules

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"database/sql"
	"fmt"

	"net"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	// Assuming these packages exist in your Go project based on the Java imports
	// "net.tc/data/db"
	// "net.tc/utils"

	"github.com/go-sql-driver/mysql"
	log "github.com/sirupsen/logrus"
	Global "testsuite/src/global"
)

// TestBase holds the configuration and state for tests.
type TestBase struct {
	ConnectionProvider           ConnectionProvider
	Config                       map[string]any
	PippoParameters              map[string]string
	Parameters                   Global.Params
	Loops                        int
	Sleep                        int
	Verbose                      bool
	Summary                      bool
	PrintConnectionTime          bool
	ReportCSV                    bool
	SchemaName                   string
	FormattingLengthForNano      int
	StartConnTimes               []int64
	EndConnTimes                 []int64
	StartTime                    float64
	EndTime                      float64
	AwsMMSessionConsistencyLevel string
}

type DataNode struct {
	Url         string
	Connection  *sql.DB
	CParms      Global.ConnectionParameters
	NodeTCPDown bool
}

func (dataNode *DataNode) Close() {
	if dataNode.Connection != nil {
		dataNode.Connection.Close()
	}
}

// NewTestBase acts as a constructor, initializing default values and maps.
func NewTestBase(params Global.Params) *TestBase {
	return &TestBase{
		Config:                  make(map[string]any),
		Parameters:              params,
		Loops:                   50,
		Sleep:                   0,
		Verbose:                 false,
		Summary:                 false,
		PrintConnectionTime:     true,
		ReportCSV:               false,
		FormattingLengthForNano: 15,
		StartConnTimes:          []int64{},
		EndConnTimes:            []int64{},
	}
}

// ExecuteLocalAction is a placeholder hook for subclasses/implementations.
func (t *TestBase) ExecuteLocalAction() {
	// Generate this in the local implementation
}

// LocalInit is a placeholder hook for subclasses/implementations.
func (t *TestBase) LocalInit() {
	// Add in your implementation specific settings
}

// Init parses arguments and configures the TestBase instance.
func (t *TestBase) Init() {

	t.Loops = t.Parameters.Loops
	t.Sleep = t.Parameters.Sleep

	// Parse booleans
	t.Verbose = t.Parameters.Verbose
	t.Summary = t.Parameters.Summary
	t.ReportCSV = t.Parameters.ReportCSV
	t.SchemaName = t.Parameters.Schema
}

// GetExecutionTime calculates the duration between start and end.
func (t *TestBase) GetExecutionTime() float64 {
	if t.StartTime > 0 && t.EndTime > 0 {
		return t.EndTime - t.StartTime
	}
	return 0
}

// ShowHelp returns a string containing usage instructions.
func (t *TestBase) ShowHelp() string {
	var sb strings.Builder
	sb.WriteString("******************************************\n")
	sb.WriteString("DB PippoParameters to use\n")
	sb.WriteString("PippoParameters are COMMA separated and the whole set must be pass as string\n")
	sb.WriteString("IE java -Xms2G -Xmx3G -classpath \"./*:./lib/*\" net.tc.testsuite.MySQLConnectionTest \"loops=10,parameters=&characterEncoding=UTF-8, url=jdbc:mysql://192.168.4.22:3306\" \n")
	sb.WriteString("url [url=jdbc:mysql://127.0.0.1:3306]\n")
	sb.WriteString("user [user=test_user]\n")
	sb.WriteString("password [password=test_password]\n")
	sb.WriteString("parameters [parameters=&useSSL=false&autoReconnect=true]\n")
	sb.WriteString("schema [schema=test]\n")
	sb.WriteString("\n*****************************************\nApplication PippoParameters \n")
	sb.WriteString("loops [loops=50]\n")
	sb.WriteString("sleep [sleep=0] value in milliseconds \n")
	sb.WriteString("verbose [verbose=false]\n")
	sb.WriteString("summary [summary=false]\n")
	sb.WriteString("reportCSV [reportCSV=false]\n")
	sb.WriteString("****************************************\n Optional \n")
	sb.WriteString("selectForceAutocommitOff [selectForceAutocommitOff=true]\n")
	return sb.String()
}

// --- Formatting Helpers ---

// FormatDF2 formats a float equivalent to Java's DecimalFormat("#,###,###,##0.00")
// Note: Go's standard library does not natively support comma separators for floats,
// so this provides standard 2-decimal formatting.
func (t *TestBase) FormatDF2(val float64) string {
	return fmt.Sprintf("%.2f", val)
}

// FormatCSV formats a float equivalent to Java's DecimalFormat("#.00")
func (t *TestBase) FormatCSV(val float64) string {
	return fmt.Sprintf("%.2f", val)
}

// Return a DataObject with an open connection
func (t *TestBase) GetConnection(cParams Global.ConnectionParameters) (bool, *DataNode) {
	attributes := "?timeout=1s"

	dataObject := &DataNode{
		CParms:      cParams,
		NodeTCPDown: false,
	}

	if cParams.UseSsl {
		if cParams.SslCertificatePath != "" {

			ca := filepath.Join(cParams.SslCertificatePath, cParams.SslCa)
			client := filepath.Join(cParams.SslCertificatePath, cParams.SslClient)
			key := filepath.Join(cParams.SslCertificatePath, cParams.SslKey)

			rootCertPool := x509.NewCertPool()

			pem, err := os.ReadFile(ca)
			if err != nil {
				log.Error(err, " While trying to connect to node (CA certificate) ", net.JoinHostPort(cParams.Host, strconv.Itoa(cParams.Port)))
				dataObject.NodeTCPDown = true
				return false, dataObject
			}
			if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
				log.Error("Failed to append PEM to pool, connecting to node ", net.JoinHostPort(cParams.Host, strconv.Itoa(cParams.Port)))
				dataObject.NodeTCPDown = true
				return false, dataObject
			}

			certs, err := tls.LoadX509KeyPair(client, key)
			if err != nil {
				log.Error(err, " While trying to connect to node (Key certificate) ", net.JoinHostPort(cParams.Host, strconv.Itoa(cParams.Port)))
				dataObject.NodeTCPDown = true
				return false, dataObject
			}

			tlsKey := fmt.Sprintf("custom_%s_%d", cParams.Host, cParams.Port)
			mysql.RegisterTLSConfig(tlsKey, &tls.Config{
				RootCAs:      rootCertPool,
				Certificates: []tls.Certificate{certs},
			})

			attributes += "&tls=" + tlsKey
		} else {
			attributes += "&tls=skip-verify"
		}
	}

	// 5. Build DSN cleanly using fmt.Sprintf
	hostPort := net.JoinHostPort(cParams.Host, strconv.Itoa(cParams.Port))
	dsn := fmt.Sprintf("%s:%s@tcp(%s)/performance_schema%s", cParams.User, cParams.Password, hostPort, attributes)

	db, err := sql.Open("mysql", dsn)
	if err != nil {
		log.Error(err.Error())
		dataObject.NodeTCPDown = true
		return false, dataObject
	}

	dataObject.Connection = db

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(cParams.PingTimeout)*time.Millisecond)
	defer cancel()

	if err := db.PingContext(ctx); err != nil {
		log.Error(err.Error())
		dataObject.NodeTCPDown = true

		dataObject.Connection.Close()

		return false, dataObject
	}

	return true, dataObject
}
